presto-main/src/main/java/com/facebook/presto/tracing/QueryStateTracingListener.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.tracing;
+
+import com.facebook.airlift.log.Logger;
+import com.facebook.presto.execution.QueryState;
+import com.facebook.presto.execution.StateMachine;
+import com.facebook.presto.spi.tracing.Tracer;
+
+import static java.util.Objects.requireNonNull;
+
+public class QueryStateTracingListener
+        implements StateMachine.StateChangeListener<QueryState>
+{
+    private static final Logger log = Logger.get(QueryStateTracingListener.class);
+    private QueryState prevState;
+    private final Tracer tracer;
+
+    public QueryStateTracingListener(Tracer tracer)
+    {
+        this.tracer = requireNonNull(tracer, "tracer is null");
+    }
+
+    @Override
+    public void stateChanged(QueryState newState)
+    {
+        synchronized (tracer) {
+            if (prevState == null) {
+                // Initial state condition
+                tracer.startBlock(newState.toString(), "");
+                prevState = newState;
+                return;
+            }
+            tracer.endBlock(prevState.toString(), "");
```

Comment:
If the previous `if` branch is to return, then we'd better break an empty line after `}` to denote it is separate logic. Same for the one below.